### PR TITLE
[4.0] Setting datetime fields to real null value for com_content

### PR DESCRIPTION
**Pull request by Hackwar**
This hopefully changes all datetime values for com_content from the pseudo-null-date 0000-00-00 00:00:00 to a real null value.

### CODE REVIEW

**Comment by richard67** [comment]
@Hackwar I don't think the database schema checker/fixer understands this in the schema update for MySQL:

> ALTER TABLE `#__content`
>     MODIFY `created` DATETIME NULL DEFAULT NULL,
>     MODIFY `modified` DATETIME NULL DEFAULT NULL,
>     MODIFY `publish_up` DATETIME NULL DEFAULT NULL,
>     MODIFY `publish_down` DATETIME NULL DEFAULT NULL;

I think it needs a single line statement for each column for the database schema checker/fixer:

> ALTER TABLE `#__content` MODIFY `created` DATETIME NULL DEFAULT NULL;
> ALTER TABLE `#__content` MODIFY `modified` DATETIME NULL DEFAULT NULL;
> ALTER TABLE `#__content` MODIFY `publish_up` DATETIME NULL DEFAULT NULL;
> ALTER TABLE `#__content` MODIFY `publish_down` DATETIME NULL DEFAULT NULL;

You could see that here:

https://github.com/joomla/joomla-cms/blob/4.0-dev/libraries/src/Schema/ChangeItem/MysqlChangeItem.php#L149-L175

**Comment by richard67** [comment]
@Hackwar Regarding my comment above: I mean that the database checker will only check the first column modification if you have all in 1 alter table statement. But when the check shows for the first column that the statement has not run yet, it will run the complete statement.

So it is formally not correct to have all in 1 statement.

On the other hand, it might be necessary that we do it like this in MySQL 8, and maybe also lower, if strict mode is on. I noticed on MySQL 8 that when running such statement for a single column in PhpMyAdmin on MySQL8, I get an SQL error complaining about the other column not having a valid default value.

If it turns out we have to do that and so betray the database checker a bit (but it would work), then we have to do that everywhere where this change for real null columns has been made already, because I've seen in schema updates of those PR that they did it like I suggested, with 1 alter table statement for every column.

**Comment by richard67** [comment]
@Hackwar When running in PhpMyAdmin on MySQL 5.7 with strict mode on, I even have to add the `checked_out_time` column to make it work, and I have to do all with 1 statement:

> ALTER TABLE `#__content`
> 	MODIFY `created` DATETIME NULL DEFAULT NULL,
> 	MODIFY `modified` DATETIME NULL DEFAULT NULL,
> 	MODIFY `checked_out_time` DATETIME NULL DEFAULT NULL,
> 	MODIFY `publish_up` DATETIME NULL DEFAULT NULL,
> 	MODIFY `publish_down` DATETIME NULL DEFAULT NULL;

**Comment by richard67** [comment]
@Hackwar I've just tested, it needs 1 statement for each column. I will make PR against your repo.

Question is: Don't you want to do the checked_out_time here, too, and close the other PR which does checked_out_time for all?

File: components/com_content/Helper/AssociationHelper.php [hunk 1]
```diff
@@ -65,13 +65,11 @@ public static function getAssociations($id = 0, $view = null, $layout = null)
...
 				{
 					// Filter by start and end dates.
-					$nullDate = $db->quote($db->getNullDate());
 					$date = Factory::getDate();
-
 					$nowDate = $db->quote($date->toSql());
 
-					$advClause[] = '(c2.publish_up = ' . $nullDate . ' OR c2.publish_up <= ' . $nowDate . ')';
-					$advClause[] = '(c2.publish_down = ' . $nullDate . ' OR c2.publish_down >= ' . $nowDate . ')';
+					$advClause[] = '(ISNULL(c2.publish_up) OR c2.publish_up <= ' . $nowDate . ')';
```
Comment (richard67) [inline components/com_content/Helper/AssociationHelper.php]: I think this `ISNULL(...)` will not work on PostgreSQL @alikon Confirmed?

Comment (richard67) [reply components/com_content/Helper/AssociationHelper.php]: Meanwhile I've researched and am sure, there is no isnull function in postgresql, it is a mysql special, so you can't use it. I've added the correction to my against your repo.

Comment (Quy) [reply components/com_content/Helper/AssociationHelper.php]: Do it like this: https://github.com/joomla/joomla-cms/pull/25362/files#diff-65f06a5ca955e1bfe518ccf5bd19bed6R184

Comment (richard67) [reply components/com_content/Helper/AssociationHelper.php]: @Quy Sure? Does it also work with real NULL values?

Comment (richard67) [reply components/com_content/Helper/AssociationHelper.php]: @Quy I see, it does work.

Comment (richard67) [reply components/com_content/Helper/AssociationHelper.php]: @Quy But there are places like the one above where ther is not really used a query object used. Should he do that there, too, and use a query object for that?

**Comment by Hackwar** [comment]
I'm going to change the ISNULL to IS NULL and see where the issue in the install SQL is. I would keep the original PR for checked_out and simply provide PRs for everything else and then we can merge everything.

**Comment by Hackwar** [comment]
I will of course do the getNullDateTime() call.

**Comment by richard67** [comment]
Which issue in the install SQL?

File: components/com_content/Helper/AssociationHelper.php (hunk 1 above)
Comment (Quy) [reply components/com_content/Helper/AssociationHelper.php]: Here is the initial PR #21901 that did this way to base off of.

**Comment by alikon** [comment]
some work (for prepared statement) has been done even if in draft wip pr #25179  maybe can help

File: components/com_content/Helper/AssociationHelper.php (hunk 1 above)
Comment (alikon) [reply components/com_content/Helper/AssociationHelper.php]: for postgresql there are 2 choices 
`SELECT (Field IS NULL) FROM ...`
or 
`COALESCE()`  which is ISO/ANSI SQL standards

File: libraries/src/Form/Field/CalendarField.php [hunk 2]
```diff
@@ -225,7 +225,7 @@ protected function getInput()
 		{
 			case 'SERVER_UTC':
 				// Convert a date to UTC based on the server timezone.
-				if ($this->value && $this->value != Factory::getDbo()->getNullDate())
+				if ($this->value && strlen($this->value) && $this->value != Factory::getDbo()->getNullDate())
```
Comment (HLeithner) [inline libraries/src/Form/Field/CalendarField.php:228]: Wouldn't this fit better?
```if ($this->value !== NULL && $this->value != Factory::getDbo()->getNullDate())```

File: libraries/src/Form/Field/CalendarField.php [hunk 3]
```diff
@@ -384,6 +384,11 @@ public function filter($value, $group = null, Registry $input = null)
 				break;
 		}
 
+		if ($return == '')
```
Comment (HLeithner) [inline libraries/src/Form/Field/CalendarField.php:387]:
```suggestion
		if ($return === '')
```

File: libraries/src/Table/Content.php [hunk 4]
```diff
@@ -247,17 +255,17 @@ public function check()
...
 		// Set publish_down to null date if not set
 		if (!$this->publish_down)
 		{
-			$this->publish_down = $this->_db->getNullDate();
+			$this->publish_down = null;
 		}
 
 		// Check the publish down date is not earlier than publish up.
-		if ($this->publish_down < $this->publish_up && $this->publish_down > $this->_db->getNullDate())
+		if (!is_null($this->publish_up) && !is_null($this->publish_up) && $this->publish_up > $this->publish_down)
```
Comment (HLeithner) [inline libraries/src/Table/Content.php:268]: I think we use === NULL not is_null in code style or?

```suggestion
		if ($this->publish_up !== NULL && $this->publish_up !== NULL && $this->publish_up > $this->publish_down)
```

Comment (richard67) [reply libraries/src/Table/Content.php:268]: @HLeithner We use both it seems, but is_null() we use by far more often.

Comment (HLeithner) [reply libraries/src/Table/Content.php:268]: iirc @mbabker changed it in the frame work from is_null to !== NULL I'm and sure if the coding style says anything about this. At least we should use the same in all places.

Comment (SharkyKZ) [reply libraries/src/Table/Content.php:268]: `is_null()` is slower. Also `null` must be in lowercase.

Comment (Hackwar) [reply libraries/src/Table/Content.php:268]: Our codestyle has no rules regarding that. If you want us to refactor to that, then I would ask to make a separate issue for that. You might want to bring that up in the ATT. 😉

File: libraries/src/Form/Field/CalendarField.php (hunk 2 above)
Comment (Hackwar) [reply libraries/src/Form/Field/CalendarField.php:228]: No, the value could be an empty string, which isn't null.

**Comment by Quy** [comment]
Replaced by #26295.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25760">issues.joomla.org/tracker/joomla-cms/25760</a>.</sub>

**Comment by richard67** [comment]
@Quy Same for PR #25715 , will be replaced by #26491 and #26490 .

**Pull request closed without merging**